README
NAME
    List::MoreUtils - Provide the stuff missing in List::Util

SYNOPSIS
        # import specific functions

        use List::MoreUtils qw(any uniq);

        if ( any { /foo/ } uniq @has_duplicates ) {
            # do stuff
        }

        # import everything

        use List::MoreUtils ':all';

        # import by API

        # has "original" any/all/none/notall behavior
        use List::MoreUtils ':like_0.22';
        # 0.22 + bsearch
        use List::MoreUtils ':like_0.24';
        # has "simplified" any/all/none/notall behavior + (n)sort_by
        use List::MoreUtils ':like_0.33';

DESCRIPTION
    List::MoreUtils provides some trivial but commonly needed functionality
    on lists which is not going to go into List::Util.

    All of the below functions are implementable in only a couple of lines
    of Perl code. Using the functions from this module however should give
    slightly better performance as everything is implemented in C. The
    pure-Perl implementation of these functions only serves as a fallback in
    case the C portions of this module couldn't be compiled on this machine.

EXPORTS
  Default behavior
    Nothing by default. To import all of this module's symbols use the
    ":all" tag. Otherwise functions can be imported by name as usual:

        use List::MoreUtils ':all';

        use List::MoreUtils qw{ any firstidx };

    Because historical changes to the API might make upgrading
    List::MoreUtils difficult for some projects, the legacy API is available
    via special import tags.

  Like version 0.22 (last release with original API)
    This API was available from 2006 to 2009, returning undef for empty
    lists on "all"/"any"/"none"/"notall":

        use List::MoreUtils ':like_0.22';

    This import tag will import all functions available as of version 0.22.
    However, it will import "any_u" as "any", "all_u" as "all", "none_u" as
    "none", and "notall_u" as "notall".

  Like version 0.24 (first incompatible change)
    This API was available from 2010 to 2011. It changed the return value of
    "none" and added the "bsearch" function.

        use List::MoreUtils ':like_0.24';

    This import tag will import all functions available as of version 0.24.
    However it will import "any_u" as "any", "all_u" as "all", and
    "notall_u" as "notall". It will import "none" as described in the
    documentation below (true for empty list).

  Like version 0.33 (second incompatible change)
    This API was available from 2011 to 2014. It is widely used in several
    CPAN modules and thus it's closest to the current API. It changed the
    return values of "any", "all", and "notall". It added the "sort_by" and
    "nsort_by" functions and the "distinct" alias for "uniq". It omitted
    "bsearch".

        use List::MoreUtils ':like_0.33';

    This import tag will import all functions available as of version 0.33.
    Note: it will not import "bsearch" for consistency with the 0.33 API.

FUNCTIONS
  Junctions
   *Treatment of an empty list*
    There are two schools of thought for how to evaluate a junction on an
    empty list:

    *   Reduction to an identity (boolean)

    *   Result is undefined (three-valued)

    In the first case, the result of the junction applied to the empty list
    is determined by a mathematical reduction to an identity depending on
    whether the underlying comparison is "or" or "and". Conceptually:

                        "any are true"      "all are true"
                        --------------      --------------
        2 elements:     A || B || 0         A && B && 1
        1 element:      A || 0              A && 1
        0 elements:     0                   1

    In the second case, three-value logic is desired, in which a junction
    applied to an empty list returns "undef" rather than true or false

    Junctions with a "_u" suffix implement three-valued logic. Those without
    are boolean.

   all BLOCK LIST
   all_u BLOCK LIST
    Returns a true value if all items in LIST meet the criterion given
    through BLOCK. Sets $_ for each item in LIST in turn:

      print "All values are non-negative"
        if all { $_ >= 0 } ($x, $y, $z);

    For an empty LIST, "all" returns true (i.e. no values failed the
    condition) and "all_u" returns "undef".

    Thus, "all_u(@list)" is equivalent to "@list ? all(@list) : undef".

    Note: because Perl treats "undef" as false, you must check the return
    value of "all_u" with "defined" or you will get the opposite result of
    what you expect.

   any BLOCK LIST
   any_u BLOCK LIST
    Returns a true value if any item in LIST meets the criterion given
    through BLOCK. Sets $_ for each item in LIST in turn:

      print "At least one non-negative value"
        if any { $_ >= 0 } ($x, $y, $z);

    For an empty LIST, "any" returns false and "any_u" returns "undef".

    Thus, "any_u(@list)" is equivalent to "@list ? any(@list) : undef".

   none BLOCK LIST
   none_u BLOCK LIST
    Logically the negation of "any". Returns a true value if no item in LIST
    meets the criterion given through BLOCK. Sets $_ for each item in LIST
    in turn:

      print "No non-negative values"
        if none { $_ >= 0 } ($x, $y, $z);

    For an empty LIST, "none" returns true (i.e. no values failed the
    condition) and "none_u" returns "undef".

    Thus, "none_u(@list)" is equivalent to "@list ? none(@list) : undef".

    Note: because Perl treats "undef" as false, you must check the return
    value of "none_u" with "defined" or you will get the opposite result of
    what you expect.

   notall BLOCK LIST
   notall_u BLOCK LIST
    Logically the negation of "all". Returns a true value if not all items
    in LIST meet the criterion given through BLOCK. Sets $_ for each item in
    LIST in turn:

      print "Not all values are non-negative"
        if notall { $_ >= 0 } ($x, $y, $z);

    For an empty LIST, "notall" returns false and "notall_u" returns
    "undef".

    Thus, "notall_u(@list)" is equivalent to "@list ? notall(@list) :
    undef".

  Transformation
   apply BLOCK LIST
    Applies BLOCK to each item in LIST and returns a list of the values
    after BLOCK has been applied. In scalar context, the last element is
    returned. This function is similar to "map" but will not modify the
    elements of the input list:

      my @list = (1 .. 4);
      my @mult = apply { $_ *= 2 } @list;
      print "\@list = @list\n";
      print "\@mult = @mult\n";
      __END__
      @list = 1 2 3 4
      @mult = 2 4 6 8

    Think of it as syntactic sugar for

      for (my @mult = @list) { $_ *= 2 }

   insert_after BLOCK VALUE LIST
    Inserts VALUE after the first item in LIST for which the criterion in
    BLOCK is true. Sets $_ for each item in LIST in turn.

      my @list = qw/This is a list/;
      insert_after { $_ eq "a" } "longer" => @list;
      print "@list";
      __END__
      This is a longer list

   insert_after_string STRING VALUE LIST
    Inserts VALUE after the first item in LIST which is equal to STRING.

      my @list = qw/This is a list/;
      insert_after_string "a", "longer" => @list;
      print "@list";
      __END__
      This is a longer list

   pairwise BLOCK ARRAY1 ARRAY2
    Evaluates BLOCK for each pair of elements in ARRAY1 and ARRAY2 and
    returns a new list consisting of BLOCK's return values. The two elements
    are set to $a and $b. Note that those two are aliases to the original
    value so changing them will modify the input arrays.

      @a = (1 .. 5);
      @b = (11 .. 15);
      @x = pairwise { $a + $b } @a, @b;     # returns 12, 14, 16, 18, 20

      # mesh with pairwise
      @a = qw/a b c/;
      @b = qw/1 2 3/;
      @x = pairwise { ($a, $b) } @a, @b;    # returns a, 1, b, 2, c, 3

   mesh ARRAY1 ARRAY2 [ ARRAY3 ... ]
   zip ARRAY1 ARRAY2 [ ARRAY3 ... ]
    Returns a list consisting of the first elements of each array, then the
    second, then the third, etc, until all arrays are exhausted.

    Examples:

      @x = qw/a b c d/;
      @y = qw/1 2 3 4/;
      @z = mesh @x, @y;         # returns a, 1, b, 2, c, 3, d, 4

      @a = ('x');
      @b = ('1', '2');
      @c = qw/zip zap zot/;
      @d = mesh @a, @b, @c;   # x, 1, zip, undef, 2, zap, undef, undef, zot

    "zip" is an alias for "mesh".

   uniq LIST
   distinct LIST
    Returns a new list by stripping duplicate values in LIST. The order of
    elements in the returned list is the same as in LIST. In scalar context,
    returns the number of unique elements in LIST.

        my @x = uniq 1, 1, 2, 2, 3, 5, 3, 4; # returns 1 2 3 5 4
        my $x = uniq 1, 1, 2, 2, 3, 5, 3, 4; # returns 5

    "distinct" is an alias for "uniq".

  Partitioning
   after BLOCK LIST
    Returns a list of the values of LIST after (and not including) the point
    where BLOCK returns a true value. Sets $_ for each element in LIST in
    turn.

      @x = after { $_ % 5 == 0 } (1..9);    # returns 6, 7, 8, 9

   after_incl BLOCK LIST
    Same as "after" but also includes the element for which BLOCK is true.

   before BLOCK LIST
    Returns a list of values of LIST up to (and not including) the point
    where BLOCK returns a true value. Sets $_ for each element in LIST in
    turn.

   before_incl BLOCK LIST
    Same as "before" but also includes the element for which BLOCK is true.

   part BLOCK LIST
    Partitions LIST based on the return value of BLOCK which denotes into
    which partition the current value is put.

    Returns a list of the partitions thusly created. Each partition created
    is a reference to an array.

      my $i = 0;
      my @part = part { $i++ % 2 } 1 .. 8;   # returns [1, 3, 5, 7], [2, 4, 6, 8]

    You can have a sparse list of partitions as well where non-set
    partitions will be undef:

      my @part = part { 2 } 1 .. 10;            # returns undef, undef, [ 1 .. 10 ]

    Be careful with negative values, though:

      my @part = part { -1 } 1 .. 10;
      __END__
      Modification of non-creatable array value attempted, subscript -1 ...

    Negative values are only ok when they refer to a partition previously
    created:

      my @idx  = ( 0, 1, -1 );
      my $i    = 0;
      my @part = part { $idx[$++ % 3] } 1 .. 8; # [1, 4, 7], [2, 3, 5, 6, 8]

  Iteration
   each_array ARRAY1 ARRAY2 ...
    Creates an array iterator to return the elements of the list of arrays
    ARRAY1, ARRAY2 throughout ARRAYn in turn. That is, the first time it is
    called, it returns the first element of each array. The next time, it
    returns the second elements. And so on, until all elements are
    exhausted.

    This is useful for looping over more than one array at once:

      my $ea = each_array(@a, @b, @c);
      while ( my ($a, $b, $c) = $ea->() )   { .... }

    The iterator returns the empty list when it reached the end of all
    arrays.

    If the iterator is passed an argument of '"index"', then it returns the
    index of the last fetched set of values, as a scalar.

   each_arrayref LIST
    Like each_array, but the arguments are references to arrays, not the
    plain arrays.

   natatime EXPR, LIST
    Creates an array iterator, for looping over an array in chunks of $n
    items at a time. (n at a time, get it?). An example is probably a better
    explanation than I could give in words.

    Example:

      my @x = ('a' .. 'g');
      my $it = natatime 3, @x;
      while (my @vals = $it->())
      {
        print "@vals\n";
      }

    This prints

      a b c
      d e f
      g

  Searching
   bsearch BLOCK LIST
    Performs a binary search on LIST which must be a sorted list of values.
    BLOCK must return a negative value if the current element (stored in $_)
    is smaller, a positive value if it is bigger and zero if it matches.

    Returns a boolean value in scalar context. In list context, it returns
    the element if it was found, otherwise the empty list.

   firstval BLOCK LIST
   first_value BLOCK LIST
    Returns the first element in LIST for which BLOCK evaluates to true.
    Each element of LIST is set to $_ in turn. Returns "undef" if no such
    element has been found.

    "first_val" is an alias for "firstval".

   lastval BLOCK LIST
   last_value BLOCK LIST
    Returns the last value in LIST for which BLOCK evaluates to true. Each
    element of LIST is set to $_ in turn. Returns "undef" if no such element
    has been found.

    "last_val" is an alias for "lastval".

   indexes BLOCK LIST
    Evaluates BLOCK for each element in LIST (assigned to $_) and returns a
    list of the indices of those elements for which BLOCK returned a true
    value. This is just like "grep" only that it returns indices instead of
    values:

      @x = indexes { $_ % 2 == 0 } (1..10);   # returns 1, 3, 5, 7, 9

   firstidx BLOCK LIST
   first_index BLOCK LIST
    Returns the index of the first element in LIST for which the criterion
    in BLOCK is true. Sets $_ for each item in LIST in turn:

      my @list = (1, 4, 3, 2, 4, 6);
      printf "item with index %i in list is 4", firstidx { $_ == 4 } @list;
      __END__
      item with index 1 in list is 4

    Returns -1 if no such item could be found.

    "first_index" is an alias for "firstidx".

   lastidx BLOCK LIST
   last_index BLOCK LIST
    Returns the index of the last element in LIST for which the criterion in
    BLOCK is true. Sets $_ for each item in LIST in turn:

      my @list = (1, 4, 3, 2, 4, 6);
      printf "item with index %i in list is 4", lastidx { $_ == 4 } @list;
      __END__
      item with index 4 in list is 4

    Returns -1 if no such item could be found.

    "last_index" is an alias for "lastidx".

  Sorting
   sort_by BLOCK LIST
    Returns the list of values sorted according to the string values
    returned by the KEYFUNC block or function. A typical use of this may be
    to sort objects according to the string value of some accessor, such as

      sort_by { $_->name } @people

    The key function is called in scalar context, being passed each value in
    turn as both $_ and the only argument in the parameters, @_. The values
    are then sorted according to string comparisons on the values returned.
    This is equivalent to

      sort { $a->name cmp $b->name } @people

    except that it guarantees the name accessor will be executed only once
    per value. One interesting use-case is to sort strings which may have
    numbers embedded in them "naturally", rather than lexically.

      sort_by { s/(\d+)/sprintf "%09d", $1/eg; $_ } @strings

    This sorts strings by generating sort keys which zero-pad the embedded
    numbers to some level (9 digits in this case), helping to ensure the
    lexical sort puts them in the correct order.

   nsort_by BLOCK LIST
    Similar to sort_by but compares its key values numerically.

  Counting and calculation
   true BLOCK LIST
    Counts the number of elements in LIST for which the criterion in BLOCK
    is true. Sets $_ for each item in LIST in turn:

      printf "%i item(s) are defined", true { defined($_) } @list;

   false BLOCK LIST
    Counts the number of elements in LIST for which the criterion in BLOCK
    is false. Sets $_ for each item in LIST in turn:

      printf "%i item(s) are not defined", false { defined($_) } @list;

   minmax LIST
    Calculates the minimum and maximum of LIST and returns a two element
    list with the first element being the minimum and the second the
    maximum. Returns the empty list if LIST was empty.

    The "minmax" algorithm differs from a naive iteration over the list
    where each element is compared to two values being the so far calculated
    min and max value in that it only requires 3n/2 - 2 comparisons. Thus it
    is the most efficient possible algorithm.

    However, the Perl implementation of it has some overhead simply due to
    the fact that there are more lines of Perl code involved. Therefore,
    LIST needs to be fairly big in order for "minmax" to win over a naive
    implementation. This limitation does not apply to the XS version.

ENVIRONMENT
    When "LIST_MOREUTILS_PP" is set, the module will always use the
    pure-Perl implementation and not the XS one. This environment variable
    is really just there for the test-suite to force testing the Perl
    implementation, and possibly for reporting of bugs. I don't see any
    reason to use it in a production environment.

MAINTENANCE
    The maintenance goal is to preserve the documented semantics of the API;
    bug fixes that bring actual behavior in line with semantics are allowed.
    New API functions may be added over time. If a backwards incompatible
    change is unavoidable, we will attempt to provide support for the legacy
    API using the same export tag mechanism currently in place.

    This module attempts to use few non-core dependencies. Non-core
    configuration and testing modules will be bundled when reasonable;
    run-time dependencies will be added only if they deliver substantial
    benefit.

BUGS
    There is a problem with a bug in 5.6.x perls. It is a syntax error to
    write things like:

        my @x = apply { s/foo/bar/ } qw{ foo bar baz };

    It has to be written as either

        my @x = apply { s/foo/bar/ } 'foo', 'bar', 'baz';

    or

        my @x = apply { s/foo/bar/ } my @dummy = qw/foo bar baz/;

    Perl 5.5.x and Perl 5.8.x don't suffer from this limitation.

    If you have a functionality that you could imagine being in this module,
    please drop me a line. This module's policy will be less strict than
    List::Util's when it comes to additions as it isn't a core module.

    When you report bugs, it would be nice if you could additionally give me
    the output of your program with the environment variable
    "LIST_MOREUTILS_PP" set to a true value. That way I know where to look
    for the problem (in XS, pure-Perl or possibly both).

SUPPORT
    Bugs should always be submitted via the CPAN bug tracker.

    You can find documentation for this module with the perldoc command.

        perldoc List::MoreUtils

    You can also look for information at:

    *   RT: CPAN's request tracker

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=List-MoreUtils>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/List-MoreUtils>

    *   CPAN Ratings

        <http://cpanratings.perl.org/l/List-MoreUtils>

    *   CPAN Search

        <http://search.cpan.org/dist/List-MoreUtils/>

    *   Git Repository

        <https://github.com/perl5-utils/List-MoreUtils>

  Where can I go for help?
    If you have a bug report, a patch or a suggestion, please open a new
    report ticket at CPAN (but please check previous reports first in case
    your issue has already been addressed) or open an issue on GitHub.

    Report tickets should contain a detailed description of the bug or
    enhancement request and at least an easily verifiable way of reproducing
    the issue or fix. Patches are always welcome, too - and it's cheap to
    send pull-requests on GitHub. Please keep in mind that code changes are
    more likely accepted when they're bundled with an approving test.

    If you think you've found a bug then please read "How to Report Bugs
    Effectively" by Simon Tatham:
    <http://www.chiark.greenend.org.uk/~sgtatham/bugs.html>.

  Where can I go for help with a concrete version?
    Bugs and feature requests are accepted against the latest version only.
    To get patches for earlier versions, you need to get an agreement with a
    developer of your choice - who may or not report the issue and a
    suggested fix upstream (depends on the license you have chosen).

  Business support and maintenance
    Generally, in volunteered projects, there is no right for support. While
    every maintainer is happy to improve the provided software, spare time
    is limited.

    For those who have a use case which requires guaranteed support, one of
    the maintainers should be hired or contracted. For business support you
    can contact Jens via his CPAN email address rehsackATcpan.org. Please
    keep in mind that business support is neither available for free nor are
    you eligible to receive any support based on the license distributed
    with this package.

THANKS
  Tassilo von Parseval
    Credits go to a number of people: Steve Purkis for giving me namespace
    advice and James Keenan and Terrence Branno for their effort of keeping
    the CPAN tidier by making List::Utils obsolete.

    Brian McCauley suggested the inclusion of apply() and provided the
    pure-Perl implementation for it.

    Eric J. Roode asked me to add all functions from his module
    "List::MoreUtil" into this one. With minor modifications, the pure-Perl
    implementations of those are by him.

    The bunch of people who almost immediately pointed out the many problems
    with the glitchy 0.07 release (Slaven Rezic, Ron Savage, CPAN testers).

    A particularly nasty memory leak was spotted by Thomas A. Lowery.

    Lars Thegler made me aware of problems with older Perl versions.

    Anno Siegel de-orphaned each_arrayref().

    David Filmer made me aware of a problem in each_arrayref that could
    ultimately lead to a segfault.

    Ricardo Signes suggested the inclusion of part() and provided the
    Perl-implementation.

    Robin Huston kindly fixed a bug in perl's MULTICALL API to make the
    XS-implementation of part() work.

  Jens Rehsack
    Credits goes to all people contributing feedback during the v0.400
    development releases.

    Special thanks goes to David Golden who spent a lot of effort to develop
    a design to support current state of CPAN as well as ancient software

Changes
Revision history for Perl extension List-MoreUtils

0.430	2020-10-21
    - bump List::MoreUtils::XS requirement to 0.430
    - release 0.429_001 without any further changes

0.429_001	2020-10-05
    - fix failing installation in parallel make (Issue#38)
    - infrastructure improvements and tooling updates,
      lots of author tests with according fixes added
    - added slide and slideatatime functions wished by SCHWERN
    - documentation fixes (PR#21, PR#33, PR#34, RT#126029, RT#132043, RT#132940)

0.428	2017-12-11
    - release 0.427_001 without further changes

0.427_001	2017-11-06
    - fix Issue#22 (Upgrading from LMU 0.416 to split XS/PP version will fail)
      reported by Graham Knop (haarg), this time hopefully the right way,
      repeated in RT#123310.
    - fix Issue#29 (pairwise() PP implementation add tail undefs if arrays
      sizes differs) reported by 0x62ash

0.426	2017-10-12
    - release 0.425_002 without further changes

0.425_002	2017-10-05
    - re-introduce Config::AutoConf
      (https://rt.cpan.org/Ticket/Display.html?id=122875#txn-1745984)
      for final satisfying RT#122875
    - bump required version of List::MoreUtils::XS to 0.426, if XS is available

0.425_001	2017-09-27
    - fix broken arrayify prototype
    - revert removal of old List::MoreUtils::XS parts

0.425	2017-09-06
    - release 0.424_001 as 0.425 with XS is META.* after 136:0 PASS:FAIL
      on CPAN Tester Matrix

0.424_001	2017-09-05
    - Makefile.PL: modify PREREQ_PM instead of recommend dynamically

0.423	2017-08-22
    - sync version with List::MoreUtils::XS
    - add examples for binsert/bremove (LMU::XS issue #1, Thanks to shawnlaffan)
    - update tests to latest List::MoreUtils::XS
    - recommend List::MoreUtils::XS 0.423

0.420_001	2017-08-15
    - add some new functions:
      * qsort (XS only)
      * binsert
      * bremove
      * listcmp
      * arrayify (RT#17230)
      * samples (RT#77562)
      * minmaxstr (RT#106401)
      * lower_bound
      * upper_bound
      * equal_range
      * frequencies
      * occurrences
      * mode (RT#91991)
      * zip6 (RT#42921)
      * reduce_0
      * reduce_1
      * reduce_u
    - improve tests
    - make List::MoreUtils::XS independent from List::MoreUtils
      Note that List::MoreUtils::XS doesn't guarantee API stability - this
      feature is only provided through List::MoreUtils as frontend.
    - improve Makefile.PL regarding some build artifacts

0.419 2017-04-06
    - Fix RT#120799: Makefile.PL fails due to unescaped paths interpolated
      in regex pattern (thanks, A. Sinan Unur)

0.418 2017-03-29
    - Clarify license conditions
    - fix Changes version/date of 0.417_001

0.417_001 2017-03-21
    - Divorce List::MoreUtils and List::MoreUtils::XS
    - Change license to Apache 2.0 to avoid code stealing without credits
    - Don't support Perl 5.6 out of the box anymore
    - fix RT#120235 - uniq examples are incorrect
    - remove things which will be never done from TODO list as suggested
      by Mark Aufflick in PR#18
    - Apply patch from Yanick Champoux to avoid spelling checker complains

0.416	2016-07-06
    - Change the way how the XS part is loaded as a result of RT#115808
    - Fix some spelling errors (fix RT#115807), thanks Salvatore Bonaccorso
    - Requires XSLoader 0.22

0.415	2016-05-01
    - Release 0.414_001 as 0.415 without further changes

0.414_001 2016-04-13
    - fix RT#75727 - after's XS implementation call XSRETURN(-1) when it
      doesn't find an element (2nd patch provided by Reini Urban,
      regression test provided by Tony Cook)
    - fix RT#113117 - XS's minmax() sometimes return undef (perl >= 5.20),
      thanks PERLANCAR and SREZIC
    - explicit test for thesis in RT#110998 - XS implementation of pairwise
      fails with memory allocation error when there are more return values
      than in original lists -- thesis is proven wrong
    - efficiency improvements by bulk88
    - improve some tests to get clearer reports
    - distinguish between "Makefile.PL find's a .git directory" and
      "Makefile.PL runs in maintainer mode"

0.413	2015-06-10
    - Fix compiling in c++ mode (depreciated, but some people seem to
      require it). Solves RT#104690

0.412	2015-05-19
    - release 0.411_001 without further changes

0.411_001	2015-05-11
    - move generation of test endpoints to author stage as requested per issue/#9
    - add a rough guide for contributors
    - fix rt#103251 to avoid removing bundled stuff by accident
    - Fix compilation errors under cl (Thanks to jddurand)

0.410	2015-03-30
    - release 0.409_003 after no further issues came up

0.409_003	2015-03-27
    - update bundled bootstrap modules
      * Data::Tumbler to 0.010
      * Test::WriteVariants to 0.012
      * Config::AutoConf to 0.311
    - fix spelling (and add stop-words for names etc. in author tests)

0.409_002	2015-03-23
    - fix multiple mg_get can break weird tie's (thanks to leont)
    - fix test run using PERL5OPT=d:Confess (thanks kentl & ribasushi)
    - use base instead of parent, cause parent isn't bundled before 5.10.1
      (smoke report from SREZIC)
    - update bundled modules (for bootstrapping) and ppport.h (from 3.25
      to 3.31)

0.409_001	2015-03-21
    - fix RT#102885: uniq bug broke tied array (reported by louying@pwrd.com)
    - fix issue/8: Macros introduced in dfd851147f cause problems with MSVC
      (reported by A. Sinan Unur)
    - Update ppport.h from 3.25 to 3.31

0.408	2015-03-18
    - fix RT#102840: uniq broken for call-by-function-return (reported by
      Jean-Damien Durand), with a new test case thanks to Thomas Sibley
    - fix RT#102853: hent_val accesses (reported by Brad Forschinger with
      a reasonable patch)
    - fix RT#102833: Compilation error with perl 5.21.7+ (reported by
      Slaven Rezic)
    - fix regex for RT#44518 test

0.407	2015-03-17
    - Added one(), onlyidx(), onlyval() (RT#73134, MHASCH) and onlyres()
    - improve XS maintainability

Makefile.PL
#!perl

use strict;
use warnings;

use 5.008_001;

use Config;
use ExtUtils::MakeMaker;

BEGIN
{
    unless (grep { $_ eq "." } @INC)
    {
        use lib ".";
    }
}

use inc::latest 'Capture::Tiny';
use inc::latest 'Config::AutoConf';

if (inc::latest->can("write"))
{
    inc::latest->write("inc");
    for my $mod (inc::latest->loaded_modules)
    {
        inc::latest->bundle_module($mod, "inc");
    }
}

use inc::Config::AutoConf::LMU ();

inc::Config::AutoConf::LMU->_set_argv(@ARGV);    # XXX hack because we cannot construct for global use
my $loadable_xs = inc::Config::AutoConf::LMU->check_produce_loadable_xs_build();

if (-d "Sandbox")
{
    unless (grep { $_ eq "." } @INC)
    {
        use lib ".";
    }
    eval "use Sandbox::Tumble ();";
    $@ and die $@;
    eval "use File::Path ();";
    File::Path->import;
    -d 't/xs'       and rmtree('t/xs');
    -d 't/pureperl' and rmtree('t/pureperl');
    Sandbox::Tumble->tumble(qw(t));
}

my $conflictMsg = <<EOCM;
***
    This version of %s conflicts with
    the version of module %s v%s you have installed.

    It's strongly recommended that you update it after
    installing this version of %s.
***
EOCM

my $selfConflictMsg = <<EOSCM;
***
    This version of %s conflicts with the version of
    %s v%s at %s you have installed.

    It's strongly recommended that you carefully read
    the Changes file and test your application with
    this version before finally upgrading.

    The installed files will be removed when installing.
***
EOSCM

my %RUN_DEPS = (
    'Exporter::Tiny' => '0.038',
    ($loadable_xs ? ('List::MoreUtils::XS' => '0.430') : ())
);
my %BUNDLE_CONFIGURE_DEPS = (
    'inc::latest'      => '0.500',
    'Config::AutoConf' => '0.315',
);
my %CONFIGURE_DEPS = (
    'ExtUtils::MakeMaker' => 0,
);
my %TEST_DEPS = (
    'Storable'   => 0,
    'Test::More' => 0.96,
);
my %CONFLICTS = (
    'List::MoreUtils'     => '0.416',
    'List::MoreUtils::PP' => '0.416',
    'List::MoreUtils::XS' => '0.416',
);

WriteMakefile1(
    META_MERGE => {
        'meta-spec' => {version => 2},
        resources   => {
            homepage   => 'https://metacpan.org/release/List-MoreUtils',
            repository => {
                url  => 'https://github.com/perl5-utils/List-MoreUtils.git',
                web  => 'https://github.com/perl5-utils/List-MoreUtils',
                type => 'git',
            },
            bugtracker => {
                web    => 'https://rt.cpan.org/Dist/Display.html?Name=List-MoreUtils',
                mailto => 'bug-List-MoreUtils@rt.cpan.org',
            },
            license => 'apache_2_0',
        },
        prereqs => {
            develop => {
                requires => {
                    %BUNDLE_CONFIGURE_DEPS,
                    'Test::CPAN::Changes'                 => 0,
                    'Test::CheckManifest'                 => 0,
                    'Module::CPANTS::Analyse'             => '0.96',
                    'Test::Kwalitee'                      => 0,
                    'Test::Perl::Critic'                  => 0,
                    'Test::PerlTidy'                      => 0,
                    'Test::Pod'                           => 0,
                    'Test::Pod::Coverage'                 => 0,
                    'Test::Pod::Spelling::CommonMistakes' => 0,
                    'Test::Spelling'                      => 0,
                    'Test::WriteVariants'                 => '0.014',
                },
            },
            configure => {
                requires   => {%CONFIGURE_DEPS},
                recommends => {'Config::AutoConf' => '0.315'}
            },
            test => {
                requires => {%TEST_DEPS},
                requires => {'Test::LeakTrace' => 0}
            },
            runtime => {requires => {%RUN_DEPS}},
        },
    },
    NAME         => 'List::MoreUtils',
    ABSTRACT     => 'Provide the stuff missing in List::Util',
    VERSION_FROM => 'lib/List/MoreUtils.pm',
    AUTHOR       => [
        'Tassilo von Parseval <tassilo.von.parseval@rwth-aachen.de>',
        'Adam Kennedy <adamk@cpan.org>',
        'Jens Rehsack <rehsack@cpan.org>'
    ],
    LICENSE            => 'apache_2_0',
    CONFIGURE_REQUIRES => \%CONFIGURE_DEPS,
    CONFLICTS          => \%CONFLICTS,
    PREREQ_PM          => \%RUN_DEPS,
    TEST_REQUIRES      => \%TEST_DEPS,
    (-d "Sandbox" ? (realclean => {FILES => "t/pureperl t/xs M*.bak"}) : ()),
    depend => {'$(FIRST_MAKEFILE)' => '$(VERSION_FROM)'},
    test   => {TESTS               => join(' ', 't/*.t', 't/pureperl/*.t', ($loadable_xs ? 't/xs/*.t' : ()), 'xt/*.t')},
    # Otherwise 'cxinc' isn't defined
    ($] < 5.012 ? (DEFINE => '-DPERL_EXT') : (),),
);

sub CheckConflicts
{
    my %params    = @_;
    my %conflicts = %{$params{CONFLICTS}};
    my $found     = 0;
    my @kill;

    while (my ($module, $version) = each(%conflicts))
    {
        undef $@;
        eval "require $module";
        next if $@;
        my $installed = eval "\$" . $module . "::VERSION";
        if ($installed le $version)
        {
            ++$found;
            if ($module eq "List::MoreUtils" or $module eq "List::MoreUtils::PP")
            {
                (my $modfile = "${module}.pm") =~ s,::,/,g;
                warn sprintf($selfConflictMsg, $params{NAME}, $module, $installed, $INC{$modfile});
                push @kill, $modfile;
            }
            else
            {
                warn sprintf($conflictMsg, $params{NAME}, $module, $installed, $params{NAME});
            }

        }
    }

    if (@kill)
    {

        package MY;
        no warnings 'once';

        *install = sub {
            my $self = shift;
            (my $targets = $self->SUPER::install) =~ s/^(pure_site_install\s+::?)/$1 delete_conflicting_files/m;
            return "\ndelete_conflicting_files :\n"
              . join("",
                map { "\t\$(NOECHO) \$(RM_F) " . $self->quote_literal($self->catfile('$(DESTINSTALLSITEARCH)', $_)) . "\n" }
                  @kill)
              . $targets;
        };
    }

    return !$found;
}

sub WriteMakefile1
{    # originally written by Alexandr Ciornii, version 0.21. Added by eumm-upgrade.
    my %params       = @_;
    my $eumm_version = $ExtUtils::MakeMaker::VERSION;
    $eumm_version = eval $eumm_version;
    die "EXTRA_META is deprecated" if (exists($params{EXTRA_META}));
    die "License not specified"    if (!exists($params{LICENSE}));
    $params{TEST_REQUIRES}
      and $eumm_version < 6.6303
      and $params{BUILD_REQUIRES} = {%{$params{BUILD_REQUIRES} || {}}, %{delete $params{TEST_REQUIRES}}};
    #EUMM 6.5502 has problems with BUILD_REQUIRES
    $params{BUILD_REQUIRES}
      and $eumm_version < 6.5503
      and $params{PREREQ_PM} = {%{$params{PREREQ_PM} || {}}, %{delete $params{BUILD_REQUIRES}}};
    ref $params{AUTHOR}
      and "ARRAY" eq ref $params{AUTHOR}
      and $eumm_version < 6.5702
      and $params{AUTHOR} = join(", ", @{$params{AUTHOR}});
    delete $params{CONFIGURE_REQUIRES}    if ($eumm_version < 6.52);
    delete $params{MIN_PERL_VERSION}      if ($eumm_version < 6.48);
    delete $params{META_MERGE}            if ($eumm_version < 6.46);
    delete $params{META_ADD}{prereqs}     if ($eumm_version < 6.58);
    delete $params{META_ADD}{'meta-spec'} if ($eumm_version < 6.58);
    delete $params{META_ADD}              if ($eumm_version < 6.46);
    delete $params{LICENSE}               if ($eumm_version < 6.31);
    delete $params{AUTHOR}                if ($] < 5.005);
    delete $params{ABSTRACT_FROM}         if ($] < 5.005);
    delete $params{BINARY_LOCATION}       if ($] < 5.005);

    # more or less taken from SQL::Statement's Makefile.PL
    if ($params{CONFLICTS})
    {
        my $ok = CheckConflicts(%params);
        exit(0) if ($params{PREREQ_FATAL} and not $ok);
        my $cpan_smoker = grep { $_ =~ m/(?:CR_SMOKER|CPAN_REPORTER|AUTOMATED_TESTING)/ } keys %ENV;
        unless ($cpan_smoker || $ENV{PERL_MM_USE_DEFAULT})
        {
            sleep 4 unless ($ok);
        }
        delete $params{CONFLICTS};
    }

    WriteMakefile(%params);
}
